src/scenic/core/vectors.py

```diff
@@ -4,427 +4,659 @@
+            return Orientation.fromEuler(*thing)
+        elif isinstance(thing, (tuple, list)):
+            if len(thing) != 3:
+                raise CoercionFailure("Cannot coerce a tuple/list of length not 3 to an orientation")
+            return Orientation.fromEuler(*thing)
+        elif canCoerceType(type(thing), float):
+            return Orientation.fromEuler(coerceToFloat(thing), 0, 0)
+        else:
+            raise CoercionFailure
+
+    @staticmethod
+    def _canCoerceType(ty):
+        if issubclass(ty, (tuple, list, Vector)):
+            return True
+        return canCoerceType(ty, float) or issubclass(ty, Orientation) or hasattr(ty, 'toOrientation')
+
+    @cached_property
+    def eulerAngles(self) -> typing.Tuple[float, float, float]:
+        """Global intrinsic Euler angles yaw, pitch, roll."""
+        # Wrapped to catch gimbal lock warning.
+        with warnings.catch_warnings():
+            warnings.simplefilter("ignore")
+            angles = self.r.as_euler('ZXY', degrees=False)
+        return angles
+
+    def getRotation(self):
+        return self.r
+
+    @cached_property
+    def inverse(self) -> Orientation:
+        return Orientation(self.r.inv())
+
+    @distributionMethod
+    def __mul__(self, other) -> Orientation:
+        if type(other) is not Orientation:
+            return NotImplemented
+        return Orientation(self.r * other.r)
+    
+    @distributionMethod
+    def __add__(self, other) -> Orientation:
+        if isinstance(other, (float, int)):
+            other = Orientation.fromEuler(other, 0, 0)
+        elif type(other) is not Orientation:
+            return NotImplemented
+        return other * self
+
+    @distributionMethod
+    def __radd__(self, other) -> Orientation:
+        if isinstance(other, (float, int)):
+            other = Orientation.fromEuler(other, 0, 0)
+        elif type(other) is not Orientation:
+            return NotImplemented
+        return self * other
+
+    def __getitem__(self, index) -> float:
+        return self.q[index]
+
+    def __repr__(self):
+        return f'Orientation.fromQuaternion({list(self.q)!r})'
+
+    def __hash__(self):
+        return hash(tuple(self.q)) + hash(tuple(-self.q))
+
+    @distributionFunction
+    def globalToLocalAngles(self, yaw, pitch, roll) -> typing.Tuple[float, float, float]:
+        """Find Euler angles w.r.t. a given parent orientation."""
+        orientation = Orientation.fromEuler(yaw, pitch, roll)
+        desiredQuat = self.inverse * orientation
+        return desiredQuat.eulerAngles
+
+    def __eq__(self, other):
+        if not isinstance(other, Orientation):
+            return NotImplemented
+        return numpy.array_equal(self.q, other.q) or numpy.array_equal(self.q, -other.q)
+
+    def approxEq(self, other, tol=1e-10):
+        if not isinstance(other, Orientation):
+            return NotImplemented
+        return abs(numpy.dot(self.q, other.q)) > 1 - tol
```

Comment:
Should also allow the dot product to be close to `-1`. (I guess this never came up in any of the tests! We should probably write one where a judicious choice of angles results in two different quaternion representations for the same orientation.)

Reply:
Doesn't the `abs` take care of this?

Reply:
Oops, you're absolutely right. Let's still add a test along the lines I suggested above though.